### PR TITLE
Add constructors for valid reconcile.Result values

### PR DESCRIPTION
**Pull request by cbandy**
The result list of Reconciler.Reconcile is a pair of types with multiple fields and special values. Some combinations are confusing, ignored, or cause warnings at runtime. The following values can be combined eighteen ways:

    Result.Requeue = { true, false }
    Result.RequeueAfter = { negative, zero, positive }
    error = { nil, non-nil, terminal }

These constructors provide names and documentation for four of the valid combinations.

**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 
 - [x] Other

### CODE REVIEW

File: internal/controller/pgupgrade/pgupgrade_controller.go [hunk 1]
```diff
@@ -493,17 +494,15 @@ func (r *PGUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
 
 	// TODO: write upgradeJob back to world? No, we will wake and see it when it
 	// has some progress. OTOH, whatever we just wrote has the latest metadata.generation.
 	// TODO: consider what it means to "re-use" the same PGUpgrade for more than
 	// one postgres version. Should the job name include the version number?
 
-	log.Info("Reconciled", "requeue", err != nil ||
-		result.Requeue ||
-		result.RequeueAfter > 0)
+	log.Info("Reconciled", "requeue", !result.IsZero() || err != nil)
```
Comment (benjaminjb) [inline internal/controller/pgupgrade/pgupgrade_controller.go:505]: 🤔 We declare `result` in the func, but where do we use it? Almost all the returns in this file are `return ctrl.Result{}, err`.

File: internal/controller/runtime/runtime.go [hunk 2]
```diff
@@ -40,6 +41,36 @@ func init() {
...
+	// When error is not nil and not a TerminalError, the controller-runtime Controller
+	// puts [reconcile.Request] back into the workqueue using AddRateLimited.
+	// - https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.4/pkg/internal/controller/controller.go#L317
+	// - https://pkg.go.dev/k8s.io/client-go/util/workqueue#RateLimitingInterface
+}
+
+// ErrorWithoutBackoff returns a Result and error that indicate a non-nil err
+// should be logged and measured without retrying its [reconcile.Request].
+// When err is nil, nothing is logged and the Request is not retried.
+func ErrorWithoutBackoff(err error) (reconcile.Result, error) {
```
[the remark this answers is not in the document]
Comment (cbandy) [reply internal/controller/runtime/runtime.go:61]: That's it.

File: internal/controller/pgupgrade/pgupgrade_controller.go (hunk 1 above)
Comment (cbandy) [reply internal/controller/pgupgrade/pgupgrade_controller.go:505]: Perhaps it's for L150, where we change `err` during a `defer`: https://github.com/CrunchyData/postgres-operator/blob/076c7ef6bc7f909e6f1c4cbc62a047155e578b7b/internal/controller/pgupgrade/pgupgrade_controller.go#L149-L150

File: internal/controller/runtime/runtime.go [hunk 3]
```diff
@@ -40,6 +41,36 @@ func init() {
 	}
 }
 
+// ErrorWithBackoff returns a Result and error that indicate a non-nil err
+// should be logged and measured and its [reconcile.Request] should be retried
+// later. When err is nil, nothing is logged and the Request is not retried.
+// When err unwraps to [reconcile.TerminalError], the Request is not retried.
+func ErrorWithBackoff(err error) (reconcile.Result, error) {
```
[the remark this answers is not in the document]
Comment (cbandy) [reply internal/controller/runtime/runtime.go]: It looks like I was trying to keep the file alphabetical. I agree, they belong together. I moved them to their own file in this package, `reconcile.go`.

**Review by dsessler7 — APPROVED** [review]
LGTM

**Comment by cbandy** [comment]
Rebased. There was a conflict in `internal/controller/postgrescluster/controller.go` imports.

**Pull request merged**